### ISSUE TITLE
[DFA F15] Macro-DFA for the dense franken (block-output taps, BP WITHIN block) + a low-LR/RAdam-class optimizer — faithful LightOn reproduction

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#158]
**Recipe-hygiene, before any fine-tuning work.** Our transformer-LM negatives used MICRO DFA (per-weight `--policy dfa`) at lr 0.001 with AdamW. LightOn's WORKING recipe (arXiv:2006.12878) is **MACRO DFA** (full BP WITHIN each transformer block, random feedback injected ONLY at block outputs) + LOW LR (5e-5) + RAdam/Adam, beta2 in {0.98,0.999}, warmup 4000. We may have tested the wrong MODE at 20x too-high LR with the wrong optimizer.

Note: LightOn's OWN LM result is ppl 52 (DFA) vs 34 (BP) — a big gap. So F15 will LIKELY still be negative on LM; its value is to CLOSE the "did we test the wrong recipe?" doubt (probably confirming our negative is honest), not to find a positive.

## Ask
1. **Macro/block-tap DFA mode** for the dense franken: a per-block DFA tap at the block OUTPUT (after the FFN residual), with full BP within the block — distinct from today's per-weight `--policy dfa`. (This is the tinydfa "DFALayer" placement: `simple`=1 tap/block at output; `full`=3 taps/block.)
2. A **low-LR-friendly optimizer** in the RAdam family — RAdam is ~100 lines (LightOn ship radam.py); OR just confirm AdamW behaves at init_lr 5e-5 + warmup 4000. "Ancient paper, modernize where it matters": keep macro + low-LR faithful, the optimizer can be modern.

## Consuming Tao experiment: F15.
Refs: LightOn 2020 arXiv:2006.12878 (macro `--dfa simple` vs micro `--dfa full`, RAdam, init_lr 5e-5, ppl 52/34).

**Comment by OriPekelman** [comment]
Shipped in f5fde23. Both asks, CPU-only per tao#18.

```
toy train franken --policy dfa,dfa --dfa-granularity block   # ask 1: macro/simple
toy train franken --optimizer radam                          # ask 2: RAdam-class
```

## How the cut is built

`llama_arch.rb` detaches the residual stream at **every block boundary** and records each block's *undetached* output; `llama_seq_engine.rb` attaches one surrogate root per block,

> `L_l = sum( tap_l ⊙ B_l·e )`, with `e` detached

whose gradient **at the tap** is exactly the random-projected output error — so ordinary autodiff then does full BP *inside* the block and stops at the cut. That is tinydfa's `DFALayer` placement (`simple`, 1 tap/block at output). The CE root stays live and trains the final norm + lm_head, which is how LightOn trains them too. The **embedding boundary is deliberately not cut**: block 0's surrogate propagates into the embedding, as a DFALayer stack does.

Note on naming: your `full` (3 taps/block) is *not* our `matmul` — ours is finer still, per weight matrix. So `matmul` (default, byte-null) = micro, `block` = your `simple`.

## Three properties, gated separately

Each can break without breaking the others, and a run that is only two of the three is a hybrid wearing a recipe's name:

1. **The forward is unchanged.** `tnn_detach` is forward-identity, so step 1 is **byte-identical to BP** (`6.464970588684082` either way). That is what makes any macro-vs-BP number comparable at all.
2. **The backward is different.** Steps 2+ diverge.
3. **The blocks really train from the injected error.** Changing `--dfa-b-seed` moves the curve — if the surrogate roots weren't reaching the weights, the blocks would be frozen and the feedback seed *couldn't* matter. This is the one that would catch a silently-unwired macro build; curve-watching would not.

Plus 6 fail-loud guards for combinations that would leave a cross-block backward path alive: `--align-events` (no chain shadow exists under macro), `--attnres` and `--mtp` (both re-open the path), `--no-shadow` (a micro-only axis), and mixed/empty policies (the cut is at every boundary, so a per-layer macro policy is not expressible — per-layer placement is the *micro* axis, toy#151).

## First numbers, and a warning about comparing arms

2-layer gate shape, 30 steps, seed 0: **BP 3.514, micro-DFA 3.541, macro-DFA 5.076, macro+RAdam 5.976.**

Micro's near-BP number is an **artifact of scope**: at the default `attn` scope micro policies only attention qkv, so most of the net is still BP, while macro policies the *entire stack*. **They are not comparable arms** — which is exactly the confusion this ticket exists to remove, and a good part of why F1 washed out. Your calibration is right: expect F15 to stay negative at scale.

## RAdam — exact where it matters, and explicitly not where it doesn't

`r_t` is a per-**step scalar**, so in the rectified regime "AdamW at lr·r_t" *is* RAdam exactly — same moments, same bias correction, no engine work, no new kernel. In the un-rectified early regime Liu et al. take a non-adaptive momentum step; **we take no step** (`r_t = 0`). At β2=0.999 that is the first 4 steps, bounded and stated rather than discovered later from a curve that doesn't match a reference implementation. Doing it properly needs a second optimizer step in the graph with per-step hp switching (the franken-moe muon idiom) — worth it only if a result ever turns on those 4 steps.

Consequence you'll hit immediately: **a 5-step smoke under `radam` looks frozen.** The runner announces `first_stepping_step=5` at startup so nobody debugs a flat curve. `--optimizer radam` also carries **β2 = 0.999** as part of its identity (ρ∞ is a function of β2, so "RAdam at β2=0.95" would rectify against a horizon nobody means); that is a real numerics change vs `adamw` and rides in run_start.

Your ask 2 alternative is also already available: `--lr 5e-5 --warmup 4000` under plain `--optimizer adamw`, and RAdam composes with warmup (the rectification multiplies the schedule, as LightOn ran it).

## CPU-only, and one trap closed

tao#18 says F15's small case gets no CUDA twin. The CUDA franken runner is *hand*-mirrored and doesn't implement macro — so `--device cuda --dfa-granularity block` would have run **micro** DFA and recorded it as macro. Rejected in the CLI **and** fails loud in the CUDA runner itself, for callers that bypass the CLI.

## Drive-by fix

The post-realize B upload in `llama_seq_engine.rb` hardcoded `nb = d_head * vocab`, ignoring the per-weight `b_douts` toy#151 introduced. Masked only because the runners call `franken_refresh_b!` before every step — but the macro taps are d_model-wide, so it would have widened the same trap. Now reads `b_douts`.

47 gate targets green (55 legs); F0 fixture and every toy#151 scope leg still byte-exact. Write-up in `docs/roadmap/dfa-arch-program-2026-08-10.md`.

**Comment by OriPekelman** [comment]
Closing: shipped in `f5fde23` — feat(dfa-arch): toy#158/F15 — macro (block-tap) DFA + a RAdam-class optimizer

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
